### PR TITLE
Add handling of invalid email addresses

### PR DESCRIPTION
**Pull request by timonegk**
When an email address is set to an invalid one (`@invalid.invalid`), a message is shown everywhere in the dashboard. Also, when loadpasswd is executed, email addresses that are reused for new accounts are set to invalid for the old user.

### CODE REVIEW

File: mafiasi/base/management/commands/loadpasswd.py [hunk 1]
```diff
@@ -53,7 +52,21 @@ def handle(self, *args, **options):
                 passwd_entry.save()
                 updated_entries += 1
 
+        # After importing the new users, unset email addresses of conflicting users
+        usernames = Mafiasi.objects.raw("""SELECT base_mafiasi.id FROM base_mafiasi
```
Comment (wichmannpas) [inline mafiasi/base/management/commands/loadpasswd.py:56]: `conflicting_usernames` would be a much clearer name.

File: mafiasi/base/management/commands/loadpasswd.py [hunk 2]
```diff
@@ -53,7 +52,21 @@ def handle(self, *args, **options):
...
                 updated_entries += 1
 
+        # After importing the new users, unset email addresses of conflicting users
+        usernames = Mafiasi.objects.raw("""SELECT base_mafiasi.id FROM base_mafiasi 
+            INNER JOIN base_passwdentry ON base_mafiasi.account = base_passwdentry.username
+            INNER JOIN base_yeargroup ON base_mafiasi.yeargroup_id = base_yeargroup.id
+            WHERE real_email IN
+                (SELECT CONCAT(username, '@informatik.uni-hamburg.de') FROM base_passwdentry)
+            AND base_yeargroup.gid <> base_passwdentry.gid
+            AND base_mafiasi.username SIMILAR TO '\d{2}\w{1,7}';""")
```
Comment (wichmannpas) [inline mafiasi/base/management/commands/loadpasswd.py:62]: You should not add an explicit `;` to the query inside code.

File: mafiasi/base/management/commands/loadpasswd.py [hunk 3]
```diff
@@ -53,7 +52,21 @@ def handle(self, *args, **options):
                 passwd_entry.save()
                 updated_entries += 1
 
+        # After importing the new users, unset email addresses of conflicting users
+        usernames = Mafiasi.objects.raw("""SELECT base_mafiasi.id FROM base_mafiasi 
+            INNER JOIN base_passwdentry ON base_mafiasi.account = base_passwdentry.username
+            INNER JOIN base_yeargroup ON base_mafiasi.yeargroup_id = base_yeargroup.id
+            WHERE real_email IN
+                (SELECT CONCAT(username, '@informatik.uni-hamburg.de') FROM base_passwdentry)
```
Comment (wichmannpas) [inline mafiasi/base/management/commands/loadpasswd.py:60]: *(non-essential)* Would the email domain rather be a configurable settings value as well?

File: mafiasi/base/management/commands/loadpasswd.py [hunk 4]
```diff
@@ -53,7 +52,21 @@ def handle(self, *args, **options):
...
+        for username in usernames:
+            user = Mafiasi.objects.get(username=username)
+            user.real_email = user.username + '@' + settings.INVALID_MAIL_DOMAIN
+            user.save()
+
         updated_entries -= new_entries # new entries were also updated
         num_in_db = PasswdEntry.objects.count()
-        print('{0} new, {1} updated, {2} processed, {3} in database'.format(
-                new_entries, updated_entries, total_entries, num_in_db))
+        print('{0} new, {1} updated, {2} processed, {3} email addresses unset, {4} in database'.format(
```
Comment (wichmannpas) [inline mafiasi/base/management/commands/loadpasswd.py:71]: Maybe we should print the names of the conflicting users to allow for manual check (or may even pass them on to the IRZ or similar).

File: mafiasi/base/management/commands/loadpasswd.py (hunk 1 above)
Comment (wichmannpas) [inline mafiasi/base/management/commands/loadpasswd.py:56]: Why selecting `.id`? The username is stored in the column ` username     | character varying(150)   |           | not null | `

**Review by wichmannpas — CHANGES_REQUESTED** [review]
This approach may not be robust: At least once in the past, an alias from a new reassigned student address to a staff account existed; that particular case was resolved by renaming the new student account rather than removing the alias to the staff account.

We probably still can apply this PR, however, we should be aware that there may occur a situation where the email stripping is undesirable.

File: mafiasi/base/management/commands/loadpasswd.py (hunk 3 above)
Comment (timonegk) [reply mafiasi/base/management/commands/loadpasswd.py:60]: I think the situation is unique enough to keep the domain here

**Comment by timonegk** [comment]
I'm sorry, I don't understand your comment. What do you mean with 'alias'?

File: mafiasi/base/management/commands/loadpasswd.py (hunk 3 above)
Comment (wichmannpas) [reply mafiasi/base/management/commands/loadpasswd.py:60]: Yeah, I agree.

**Comment by wichmannpas** [comment]
> I'm sorry, I don't understand your comment. What do you mean with 'alias'?

student Nabc migrated to staff account abc, but kept an email alias `0Nabc -> abc`.
10 years later, a new student was assigned the new account `Nabc` (accidentally, I am unsure whether measures have been taken to prevent this by now – if so, this edge case can not occur anymore). Due to the alias, that new account was not able to receive emails under that account.
To fix that problem, the new student account was recreated with a different name (e.g., adding the first letter of the first name to it).
-> As a consequence, the ten year old Mafiasi account would still need to keep the previous email, even though it was temporarily conflicting.

Anyways, as said, this edge case should be that rare that we can ignore it, as long as we keep in mind that it may occur.

**Review by wichmannpas — APPROVED** [review]

**Comment by timonegk** [comment]
Okay, thanks for clearing that up. I agree that we can ignore this very specific case but keep it in mind.

**Pull request merged**